### ISSUE TITLE
IAsyncResult not signaled in case of Session Error

### ISSUE THREAD

**Issue by EricGarnier** [sshnet/SSH.NET#562]
The IAsyncResult returned by Command.BeginExecute is not signaled in case of session error. Like for exemple a remote host rebooting during the command execution.
This prevent the use of .net async/await pattern like
                            await Task.Factory.FromAsync(command.BeginExecute, command.EndExecute, null);

To reproduce 👍 again a linux host, try this
                using (var sshClient = new SshClient(connectionInfo))
                {
                    sshClient.Connect();
                    using (var command = sshClient.CreateCommand(@"sleep 100000"))
                    {
                        command.CommandTimeout = TimeSpan.FromMinutes(1);
                        try
                        {
                            await Task.Factory.FromAsync(command.BeginExecute, command.EndExecute, null);
                        }
                        catch (Exception ex2)
                        {
                            Console.Write("Exception 1");
                            Console.Write(ex2.ToString());
                        }
                    }
                }

While the program wait in the sleep, reboot the host.
The exception is never throw.

If you repeat this with a 
                            command.Execute();
an exception in throw.

My analyze is that the wait handle used for session exception is not the same than the one in the IAsyncResult.

**Comment by patgauvingeek** [comment]
Salut Éric,

Even without rebooting, I can't have it to throw a `SshOperationTimeoutException` with `FromAsync`.

I used `StartNew`with `SshCommand.Execute` to fix it. The `TaskCreationOptions.LongRunning` prevents a threadpool starvation situation. I don't think this option is required in all situations.

```
private static Task<String> ExecuteAsync(this SshCommand command)
{
  return Task<String>.Factory.StartNew(command.Execute, TaskCreationOptions.LongRunning);
}
```

I would really like having async/await support on SSH.Net.

**Comment by EricGarnier** [comment]
Salut Patrick
Nice workaround.
I already finish using the synchronous version, but your wrapper is clearly better.

**Comment by Rob-Hague** [comment]
Hi @EricGarnier, I also noticed this when working on a different problem and fixed it at the same time (#1423). It did not cross my mind that the issue had long been addressed already. I'm sorry that you went through 2 PRs without any response. I will close them out now.
